p2p/transport/webrtc/state.go

```diff
@@ -0,0 +1,59 @@
+package libp2pwebrtc
+
+import (
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+type channelState uint8
+
+const (
+	stateOpen channelState = iota
+	stateReadClosed
+	stateWriteClosed
+	stateClosed
+)
+
+func (state channelState) handleIncomingFlag(flag pb.Message_Flag) channelState {
+	if state == stateClosed {
+		return state
+	}
+	switch flag {
+	case pb.Message_FIN:
+		if state == stateWriteClosed {
+			return stateClosed
+		}
+		return stateReadClosed
+
+	case pb.Message_STOP_SENDING:
+		if state == stateReadClosed {
+			return stateClosed
+		}
+		return stateWriteClosed
+	case pb.Message_RESET:
+		return stateClosed
+
+	}
+	return state
+}
+
+func (state channelState) processOutgoingFlag(flag pb.Message_Flag) channelState {
+	if state == stateClosed {
+		return state
+	}
+
+	switch flag {
+	case pb.Message_FIN:
```

Comment:
a simple mapping might make this code faster and simpler

Reply:
there are only 3 flags, so this should run in O(1) with a low constant overhead. What would we save with a mapping?

Reply:
With faster I do not mean more performant, sorry for the bad wording. Meant more like faster to dev and use.
In Go I prefer usually a mapping, as it's less to types and maps are fast enough for Go, unless it really matters. Where it does matter you could often go for an array approach as well, which is similar to a mapping but with your indexes being the thing (even for enum to strs that's an option if you take offset into account).

My biggest beef with switch statements in Go is that they give no guarantees. So you def. do want to check the default case. Second bad thing in Go is that enums also give low guarantees, as such your enum is not really an enum and can be safely casted from any int-like value to your enum, even if that's not a valid variant. So that's a second reason why the default case brings some sanity there (even if it means a panic).

Reply:
I don't want to use a DAT for this, since a bad flag could break it. I have added a default case for this function.

Reply:
fine for me to close it.